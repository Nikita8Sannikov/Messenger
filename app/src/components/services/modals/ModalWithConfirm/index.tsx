import { useState, useEffect, memo } from "react";
import { useTranslation } from "react-i18next";
import Modal from "@mui/material/Modal";
import Box from "@mui/material/Box";
import Typography from "@mui/material/Typography";
import Button from "@mui/material/Button";

import { useAppDispatch, useAppSelector } from "../../../../hooks/useGlobalState";
import { selectMainState, setModalConfirm } from "../../../../store/main/slice";

import "./modal-with-confirm.scss";

const modalTitle = "modal-confirm-title";
const modalDescription = "modal-confirm-description";

export default memo(function ModalWithConfirm() {
    const [open, setOpen] = useState(false);
    
    const { t } = useTranslation();
    const { modalConfirm } = useAppSelector(selectMainState);
    const dispatch = useAppDispatch();

    useEffect(() => {
        setOpen(Boolean(modalConfirm && modalConfirm.text));
    }, [modalConfirm]);

    // Выполнение действия
    const onAction = () => {
        if (modalConfirm && modalConfirm.cb) {
            modalConfirm.cb();
            dispatch(setModalConfirm(null));
        }
    };

    // Закрытие модального окна
    const onClose = () => {
        dispatch(setModalConfirm(null));
        setOpen(false);
    };

    return <Modal open={open} onClose={onClose} aria-labelledby={modalTitle} aria-describedby={modalDescription}>
        <Box className="modal-confirm-container">
            <Typography id={modalTitle} variant="subtitle1" component="h2">
                {modalConfirm && modalConfirm.text}
            </Typography>

            <Typography id={modalDescription} className="modal-confirm-container__buttons">
                <Button size="small" variant="outlined" color="primary" onClick={onAction}>{modalConfirm && modalConfirm.btnActionTitle}</Button>
                <Button size="small" variant="outlined" color="error" onClick={onClose}>{ t("modals.cancel") }</Button>
            </Typography>
        </Box>
    </Modal>
});