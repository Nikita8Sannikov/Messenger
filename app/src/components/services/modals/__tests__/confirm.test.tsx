import { fireEvent, render, screen } from "@testing-library/react";

import ModalWithConfirm from "@components/services/modals/confirm";
import mockUIStore from "../../../../__mocks__/@store/ui";

describe("ModalWithConfirm", () => {
	beforeEach(() => {
		jest.clearAllMocks();
	});

	it("should not render when confirmModal is null", () => {
		const store = mockUIStore.getState();
		store.confirmModal = null;

		const { container } = render(<ModalWithConfirm />);
		expect(container).toBeEmptyDOMElement();
		expect(container).toMatchSnapshot();
	});

	it("should render with text", () => {
		const store = mockUIStore.getState();
		store.confirmModal = {
			text: "Test confirmation",
			btnActionTitle: "Confirm",
			cb: jest.fn(),
		};

		const { container } = render(<ModalWithConfirm />);
        
		expect(screen.getByText("Test confirmation")).toBeInTheDocument();
		expect(screen.getByText("Confirm")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("should render with custom button text", () => {
		const store = mockUIStore.getState();
		store.confirmModal = {
			text: "Test text",
			btnActionTitle: "Test",
			cb: jest.fn(),
		};

		const { container } = render(<ModalWithConfirm />);
        
		expect(screen.getByText("Test text")).toBeInTheDocument();
		expect(screen.getByText("Test")).toBeInTheDocument();
		expect(container).toMatchSnapshot();
	});

	it("should call callback and close modal when confirm button is clicked", async () => {
		const store = mockUIStore.getState();
		const mockCallback = jest.fn();
		store.confirmModal = {
			text: "Test confirmation",
			btnActionTitle: "Confirm",
			cb: mockCallback,
		};

		const { container } = render(<ModalWithConfirm />);
        
		const confirmButton = screen.getByText("Confirm");
		await fireEvent.click(confirmButton);

		await new Promise(resolve => setTimeout(resolve, 0));

		expect(mockCallback).toHaveBeenCalled();
		expect(store.setConfirmModal).toHaveBeenCalledWith(null);
		expect(container).toMatchSnapshot();
	});

	it("should close modal when cancel button is clicked", () => {
		const store = mockUIStore.getState();
		store.confirmModal = {
			text: "Test confirmation",
			btnActionTitle: "Test",
			cb: jest.fn(),
		};

		const { container } = render(<ModalWithConfirm />);
        
		const cancelButton = screen.getByText("Cancel");
		fireEvent.click(cancelButton);

		expect(store.setConfirmModal).toHaveBeenCalledWith(null);
		expect(container).toMatchSnapshot();
	});

	it("should not call callback if it is not provided", async () => {
		const store = mockUIStore.getState();
		store.confirmModal = {
			text: "Test confirmation",
			btnActionTitle: "Confirm",
			cb: () => {},
		};

		const { container } = render(<ModalWithConfirm />);
        
		const confirmButton = screen.getByText("Confirm");
		await fireEvent.click(confirmButton);

		await new Promise(resolve => setTimeout(resolve, 0));

		expect(store.setConfirmModal).toHaveBeenCalledWith(null);
		expect(container).toMatchSnapshot();
	});
});
