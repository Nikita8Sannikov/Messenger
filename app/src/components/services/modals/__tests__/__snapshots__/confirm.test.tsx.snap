// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ModalWithConfirm should call callback and close modal when confirm button is clicked 1`] = `<div />`;

exports[`ModalWithConfirm should close modal when cancel button is clicked 1`] = `<div />`;

exports[`ModalWithConfirm should not call callback if it is not provided 1`] = `<div />`;

exports[`ModalWithConfirm should not render when confirmModal is null 1`] = `<div />`;

exports[`ModalWithConfirm should render with custom button text 1`] = `<div />`;

exports[`ModalWithConfirm should render with text 1`] = `<div />`;
